helper/restore_helper.go

```
@@ -17,14 +20,70 @@ import (
 /*
  * Restore specific functions
  */
+type ReaderType string
+const (
+	SEEKABLE ReaderType = "seekable"	// reader which supports seek
+	NONSEEKABLE			= "discard"		// reader which is not seekable
+	SUBSET				= "subset"		// reader which operates on pre filtered data
+)
+
+/* RestoreReader structure to wrap the underlying reader.
+ * readerType identifies how the reader can be used
+ * SEEKABLE uses seekReader. Used when restoring from uncompressed data with filters from local filesystem
+ * NONSEEKABLE and SUBSET types uses bufReader.
+ * SUBSET type applies when restoring using plugin(if compatible) from uncompressed data with filters
+ * NONSEEKABLE type applies for every other restore scenario
+ */
+type RestoreReader struct {
+	bufReader  *bufio.Reader
+	seekReader io.ReadSeeker
+	readerType ReaderType
+}
+
+func (r *RestoreReader) positionReader(pos uint64) error {
+	switch r.readerType {
+	case SEEKABLE:
+		seekPosition, err := r.seekReader.Seek(int64(pos), io.SeekCurrent)
+		if err != nil {
+			// Always hard quit if data reader has issues
+			_ = utils.RemoveFileIfExists(currentPipe)
+			return err
+		}
+		log(fmt.Sprintf("Data Reader seeked forward to %d byte offset", seekPosition))
+	case NONSEEKABLE:
+		numDiscarded, err := r.bufReader.Discard(int(pos))
+		if err != nil {
+			// Always hard quit if data reader has issues
+			_ = utils.RemoveFileIfExists(currentPipe)
```

nit: lines 49 and 57 ignores errors if the files fail to get cleaned up.